jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/ServerWebSocketContainer.java

```diff
@@ -226,24 +227,31 @@ public void addMapping(PathSpec pathSpec, WebSocketCreator creator)
         if (mappings.getWebSocketNegotiator(pathSpec) != null)
             throw new WebSocketException("Duplicate WebSocket Mapping for PathSpec " + pathSpec);
 
-        org.eclipse.jetty.websocket.core.server.WebSocketCreator coreCreator = (request, response, callback) ->
-        {
-            try
-            {
-                Object webSocket = creator.createWebSocket(new ServerUpgradeRequestDelegate(request), new ServerUpgradeResponseDelegate(request, response), callback);
-                if (webSocket == null)
-                    callback.succeeded();
-                return webSocket;
-            }
-            catch (Throwable x)
-            {
-                callback.failed(x);
-                return null;
-            }
-        };
+        var coreCreator = newWebSocketCreator(creator);
         mappings.addMapping(pathSpec, coreCreator, factory, configuration);
     }
 
+    /**
+     * <p>Matches the given {@code request} against existing WebSocket mappings,
+     * upgrading to WebSocket if there is a match.</p>
+     * <p>Direct upgrades without using WebSocket mappings may be performed via
+     * {@link #upgrade(WebSocketCreator, Request, Response, Callback)}.</p>
+     * <p>When {@code true} is returned, a response has been sent to the client
+     * and the {@code callback} has been completed; either because of a successful
+     * WebSocket upgrade, or because an error has occurred.</p>
+     * <p>When {@code false} is returned, a response has not been sent to the
+     * client, and the {@code callback} has not been completed; typically because
+     * the request path does not match any existing WebSocket mappings, so that
+     * the request can be handled by other {@link Handler}s.</p>
+     *
+     * @param request the request to handle, possibly a WebSocket upgrade request
+     * @param response the response to handle
+     * @param callback the callback to complete when the handling is complete
+     * @return {@code true} in case of WebSocket upgrades or failures,
+     * {@code false} if the request was not handled
+     * @see #addMapping(PathSpec, WebSocketCreator)
+     * @see #upgrade(WebSocketCreator, Request, Response, Callback)
+     */
     public boolean handle(Request request, Response response, Callback callback)
     {
         String target = Request.getPathInContext(request);
```

Comment:
Not really applicable to this PR, but I don't like the line below.  We are always creating the lambda that captures request on every single request, even if it is not an upgrade.   Can't we look up the mappings first and only start allocating stuff if there is one???

Reply:
@gregw I fixed that by using a `BiConsumer` and a static method that does not capture.